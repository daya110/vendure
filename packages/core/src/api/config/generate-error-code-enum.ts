import { buildSchema, extendSchema, GraphQLSchema, parse } from 'graphql';

export const ERROR_INTERFACE_NAME = 'ErrorResult';

/**
 * Generates the members of the `ErrorCode` enum dynamically, by getting the names of
 * all the types which inherit from the `ErrorResult` interface.
 */
export function generateErrorCodeEnum(typeDefsOrSchema: string | GraphQLSchema): GraphQLSchema {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;
    const errorNodes = Object.values(schema.getTypeMap())
        .map(type => type.astNode)
        .filter(node => {
            return (
                node &&
                node?.kind === 'ObjectTypeDefinition' &&
                node.interfaces?.map(i => i.name.value).includes(ERROR_INTERFACE_NAME)
            );
        });
    if (!errorNodes.length) {
        return schema;
    }

    const errorCodeEnum = `
        extend enum ErrorCode {
            ${errorNodes.map(n => camelToUpperSnakeCase(n?.name.value || '')).join('\n')}
        }`;
    return extendSchema(schema, parse(errorCodeEnum));
}

function camelToUpperSnakeCase(input: string): string {
    return input.replace(/([a-z])([A-Z])/g, '$1_$2').toUpperCase();
}
