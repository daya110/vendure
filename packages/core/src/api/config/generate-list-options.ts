import { stitchSchemas } from '@graphql-tools/stitch';
import {
    buildSchema,
    GraphQLEnumType,
    GraphQLField,
    GraphQLInputFieldConfig,
    GraphQLInputFieldConfigMap,
    GraphQLInputObjectType,
    GraphQLInputType,
    GraphQLInt,
    GraphQLNamedType,
    GraphQLObjectType,
    GraphQLOutputType,
    GraphQLSchema,
    isEnumType,
    isListType,
    isNonNullType,
    isObjectType,
} from 'graphql';

/**
 * Generates ListOptions inputs for queries which return PaginatedList types.
 */
export function generateListOptions(typeDefsOrSchema: string | GraphQLSchema): GraphQLSchema {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;
    const queryType = schema.getQueryType();
    if (!queryType) {
        return schema;
    }
    const objectTypes = Object.values(schema.getTypeMap()).filter(isObjectType);
    const allFields = objectTypes.reduce((fields, type) => {
        const typeFields = Object.values(type.getFields()).filter(f => isListQueryType(f.type));
        return [...fields, ...typeFields];
    }, [] as Array<GraphQLField<any, any>>);
    const generatedTypes: GraphQLNamedType[] = [];

    for (const query of allFields) {
        const targetTypeName = unwrapNonNullType(query.type).toString().replace(/List$/, '');
        const targetType = schema.getType(targetTypeName);
        if (targetType && isObjectType(targetType)) {
            const sortParameter = createSortParameter(schema, targetType);
            const filterParameter = createFilterParameter(schema, targetType);
            const existingListOptions = schema.getType(
                `${targetTypeName}ListOptions`,
            ) as GraphQLInputObjectType | null;
            const generatedListOptions = new GraphQLInputObjectType({
                name: `${targetTypeName}ListOptions`,
                fields: {
                    skip: { type: GraphQLInt },
                    take: { type: GraphQLInt },
                    sort: { type: sortParameter },
                    filter: { type: filterParameter },
                    ...(existingListOptions ? existingListOptions.getFields() : {}),
                },
            });

            if (!query.args.find(a => a.type.toString() === `${targetTypeName}ListOptions`)) {
                query.args.push({
                    name: 'options',
                    type: generatedListOptions,
                    description: null,
                    defaultValue: null,
                    extensions: null,
                    astNode: null,
                });
            }

            generatedTypes.push(filterParameter);
            generatedTypes.push(sortParameter);
            generatedTypes.push(generatedListOptions);
        }
    }
    return stitchSchemas({ schemas: [schema, generatedTypes] });
}

function isListQueryType(type: GraphQLOutputType): type is GraphQLObjectType {
    const innerType = unwrapNonNullType(type);
    return isObjectType(innerType) && !!innerType.getInterfaces().find(i => i.name === 'PaginatedList');
}

function createSortParameter(schema: GraphQLSchema, targetType: GraphQLObjectType) {
    const fields = Object.values(targetType.getFields());
    const targetTypeName = targetType.name;
    const SortOrder = schema.getType('SortOrder') as GraphQLEnumType;

    const sortableTypes = ['ID', 'String', 'Int', 'Float', 'DateTime'];
    return new GraphQLInputObjectType({
        name: `${targetTypeName}SortParameter`,
        fields: fields
            .filter(field => sortableTypes.includes(unwrapNonNullType(field.type).name))
            .reduce((result, field) => {
                const fieldConfig: GraphQLInputFieldConfig = {
                    type: SortOrder,
                };
                return {
                    ...result,
                    [field.name]: fieldConfig,
                };
            }, {} as GraphQLInputFieldConfigMap),
    });
}

function createFilterParameter(schema: GraphQLSchema, targetType: GraphQLObjectType): GraphQLInputObjectType {
    const fields = Object.values(targetType.getFields());
    const targetTypeName = targetType.name;
    const { StringOperators, BooleanOperators, NumberOperators, DateOperators } = getCommonTypes(schema);

    return new GraphQLInputObjectType({
        name: `${targetTypeName}FilterParameter`,
        fields: fields.reduce((result, field) => {
            const filterType = getFilterType(field);
            if (!filterType) {
                return result;
            }
            const fieldConfig: GraphQLInputFieldConfig = {
                type: filterType,
            };
            return {
                ...result,
                [field.name]: fieldConfig,
            };
        }, {} as GraphQLInputFieldConfigMap),
    });

    function getFilterType(field: GraphQLField<any, any>): GraphQLInputType | undefined {
        if (isListType(field.type)) {
            return;
        }
        const innerType = unwrapNonNullType(field.type);
        if (isEnumType(innerType)) {
            return StringOperators;
        }
        switch (innerType.name) {
            case 'String':
                return StringOperators;
            case 'Boolean':
                return BooleanOperators;
            case 'Int':
            case 'Float':
                return NumberOperators;
            case 'DateTime':
                return DateOperators;
            default:
                return;
        }
    }
}

function getCommonTypes(schema: GraphQLSchema) {
    const SortOrder = schema.getType('SortOrder') as GraphQLEnumType | null;
    const StringOperators = schema.getType('StringOperators') as GraphQLInputType | null;
    const BooleanOperators = schema.getType('BooleanOperators') as GraphQLInputType | null;
    const NumberRange = schema.getType('NumberRange') as GraphQLInputType | null;
    const NumberOperators = schema.getType('NumberOperators') as GraphQLInputType | null;
    const DateRange = schema.getType('DateRange') as GraphQLInputType | null;
    const DateOperators = schema.getType('DateOperators') as GraphQLInputType | null;
    if (
        !SortOrder ||
        !StringOperators ||
        !BooleanOperators ||
        !NumberRange ||
        !NumberOperators ||
        !DateRange ||
        !DateOperators
    ) {
        throw new Error(`A common type was not defined`);
    }
    return {
        SortOrder,
        StringOperators,
        BooleanOperators,
        NumberOperators,
        DateOperators,
    };
}

/**
 * Unwraps the inner type if it is inside a non-nullable type
 */
function unwrapNonNullType(type: GraphQLOutputType): GraphQLNamedType {
    if (isNonNullType(type)) {
        return type.ofType;
    }
    return type;
}
