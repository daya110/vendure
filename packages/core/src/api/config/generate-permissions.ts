import { stitchSchemas, ValidationLevel } from '@graphql-tools/stitch';
import { GraphQLEnumType, GraphQLSchema } from 'graphql';
import { GraphQLEnumValueConfigMap } from 'graphql/type/definition';

import { getAllPermissionsMetadata } from '../../common/constants';
import { PermissionDefinition } from '../../common/permission-definition';

const PERMISSION_DESCRIPTION = `@description
Permissions for administrators and customers. Used to control access to
GraphQL resolvers via the {@link Allow} decorator.

## Understanding Permission.Owner

\`Permission.Owner\` is a special permission which is used in some of the Vendure resolvers to indicate that that resolver should only
be accessible to the "owner" of that resource.

For example, the Shop API \`activeCustomer\` query resolver should only return the Customer object for the "owner" of that Customer, i.e.
based on the activeUserId of the current session. As a result, the resolver code looks like this:

@example
\`\`\`TypeScript
\\@Query()
\\@Allow(Permission.Owner)
async activeCustomer(\\@Ctx() ctx: RequestContext): Promise<Customer | undefined> {
  const userId = ctx.activeUserId;
  if (userId) {
    return this.customerService.findOneByUserId(ctx, userId);
  }
}
\`\`\`

Here we can see that the "ownership" must be enforced by custom logic inside the resolver. Since "ownership" cannot be defined generally
nor statically encoded at build-time, any resolvers using \`Permission.Owner\` **must** include logic to enforce that only the owner
of the resource has access. If not, then it is the equivalent of using \`Permission.Public\`.


@docsCategory common`;

/**
 * Generates the `Permission` GraphQL enum based on the default & custom permission definitions.
 */
export function generatePermissionEnum(
    schema: GraphQLSchema,
    customPermissions: PermissionDefinition[],
): GraphQLSchema {
    const allPermissionsMetadata = getAllPermissionsMetadata(customPermissions);
    const values: GraphQLEnumValueConfigMap = {};
    let i = 0;
    for (const entry of allPermissionsMetadata) {
        values[entry.name] = {
            value: i,
            description: entry.description,
        };
        i++;
    }

    const permissionsEnum = new GraphQLEnumType({
        name: 'Permission',
        description: PERMISSION_DESCRIPTION,
        values,
    });

    return stitchSchemas({
        subschemas: [schema],
        types: [permissionsEnum],
        typeMergingOptions: { validationSettings: { validationLevel: ValidationLevel.Off } },
    });
}
