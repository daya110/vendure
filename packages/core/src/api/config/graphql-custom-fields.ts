import { CustomFieldType } from '@vendure/common/lib/shared-types';
import { assertNever } from '@vendure/common/lib/shared-utils';
import { buildSchema, extendSchema, GraphQLInputObjectType, GraphQLSchema, parse } from 'graphql';

import { CustomFieldConfig, CustomFields } from '../../config/custom-field/custom-field-types';

/**
 * Given a CustomFields config object, generates an SDL string extending the built-in
 * types with a customFields property for all entities, translations and inputs for which
 * custom fields are defined.
 */
export function addGraphQLCustomFields(
    typeDefsOrSchema: string | GraphQLSchema,
    customFieldConfig: CustomFields,
    publicOnly: boolean,
): GraphQLSchema {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;

    let customFieldTypeDefs = '';

    if (!schema.getType('JSON')) {
        customFieldTypeDefs += `
            scalar JSON
        `;
    }

    if (!schema.getType('DateTime')) {
        customFieldTypeDefs += `
            scalar DateTime
        `;
    }

    for (const entityName of Object.keys(customFieldConfig)) {
        const customEntityFields = (customFieldConfig[entityName as keyof CustomFields] || []).filter(
            config => {
                return !config.internal && (publicOnly === true ? config.public !== false : true);
            },
        );

        const localeStringFields = customEntityFields.filter(field => field.type === 'localeString');
        const nonLocaleStringFields = customEntityFields.filter(field => field.type !== 'localeString');
        const writeableLocaleStringFields = localeStringFields.filter(field => !field.readonly);
        const writeableNonLocaleStringFields = nonLocaleStringFields.filter(field => !field.readonly);

        if (schema.getType(entityName)) {
            if (customEntityFields.length) {
                customFieldTypeDefs += `
                    type ${entityName}CustomFields {
                        ${mapToFields(customEntityFields, getGraphQlType)}
                    }

                    extend type ${entityName} {
                        customFields: ${entityName}CustomFields
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend type ${entityName} {
                        customFields: JSON
                    }
                `;
            }
        }

        if (localeStringFields.length && schema.getType(`${entityName}Translation`)) {
            customFieldTypeDefs += `
                    type ${entityName}TranslationCustomFields {
                         ${mapToFields(localeStringFields, getGraphQlType)}
                    }

                    extend type ${entityName}Translation {
                        customFields: ${entityName}TranslationCustomFields
                    }
                `;
        }

        if (schema.getType(`Create${entityName}Input`)) {
            if (writeableNonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Create${entityName}CustomFieldsInput {
                       ${mapToFields(writeableNonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Create${entityName}Input {
                        customFields: Create${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                   extend input Create${entityName}Input {
                       customFields: JSON
                   }
               `;
            }
        }

        if (schema.getType(`Update${entityName}Input`)) {
            if (writeableNonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Update${entityName}CustomFieldsInput {
                       ${mapToFields(writeableNonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Update${entityName}Input {
                        customFields: Update${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input Update${entityName}Input {
                        customFields: JSON
                    }
                `;
            }
        }

        if (customEntityFields.length && schema.getType(`${entityName}SortParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}SortParameter {
                         ${mapToFields(customEntityFields, () => 'SortOrder')}
                    }
                `;
        }

        if (customEntityFields.length && schema.getType(`${entityName}FilterParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}FilterParameter {
                         ${mapToFields(customEntityFields, getFilterOperator)}
                    }
                `;
        }

        if (writeableLocaleStringFields && schema.getType(`${entityName}TranslationInput`)) {
            if (writeableLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input ${entityName}TranslationCustomFieldsInput {
                        ${mapToFields(writeableLocaleStringFields, getGraphQlType)}
                    }

                    extend input ${entityName}TranslationInput {
                        customFields: ${entityName}TranslationCustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input ${entityName}TranslationInput {
                        customFields: JSON
                    }
                `;
            }
        }
    }

    return extendSchema(schema, parse(customFieldTypeDefs));
}

export function addServerConfigCustomFields(
    typeDefsOrSchema: string | GraphQLSchema,
    customFieldConfig: CustomFields,
) {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;
    const customFieldTypeDefs = `
            type CustomFields {
                ${Object.keys(customFieldConfig).reduce(
                    (output, name) => output + name + `: [CustomFieldConfig!]!\n`,
                    '',
                )}
            }

            extend type ServerConfig {
                customFieldConfig: CustomFields!
            }
        `;

    return extendSchema(schema, parse(customFieldTypeDefs));
}

/**
 * If CustomFields are defined on the OrderLine entity, then an extra `customFields` argument
 * must be added to the `addItemToOrder` and `adjustOrderLine` mutations.
 */
export function addOrderLineCustomFieldsInput(
    typeDefsOrSchema: string | GraphQLSchema,
    orderLineCustomFields: CustomFieldConfig[],
): GraphQLSchema {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;
    if (!orderLineCustomFields || orderLineCustomFields.length === 0) {
        return schema;
    }
    const schemaConfig = schema.toConfig();
    const mutationType = schemaConfig.mutation;
    if (!mutationType) {
        return schema;
    }
    const addItemToOrderMutation = mutationType.getFields().addItemToOrder;
    const adjustOrderLineMutation = mutationType.getFields().adjustOrderLine;
    if (!addItemToOrderMutation || !adjustOrderLineMutation) {
        return schema;
    }
    const input = new GraphQLInputObjectType({
        name: 'OrderLineCustomFieldsInput',
        fields: orderLineCustomFields.reduce((fields, field) => {
            return { ...fields, [field.name]: { type: schema.getType(getGraphQlType(field.type)) } };
        }, {}),
    });

    schemaConfig.types.push(input);
    addItemToOrderMutation.args.push({
        name: 'customFields',
        type: input,
        description: null,
        defaultValue: null,
        extensions: null,
        astNode: null,
    });
    adjustOrderLineMutation.args.push({
        name: 'customFields',
        type: input,
        description: null,
        defaultValue: null,
        extensions: null,
        astNode: null,
    });

    return new GraphQLSchema(schemaConfig);
}

type GraphQLFieldType = 'DateTime' | 'String' | 'Int' | 'Float' | 'Boolean' | 'ID';

/**
 * Maps an array of CustomFieldConfig objects into a string of SDL fields.
 */
function mapToFields(fieldDefs: CustomFieldConfig[], typeFn: (fieldType: CustomFieldType) => string): string {
    return fieldDefs.map(field => `${field.name}: ${typeFn(field.type)}`).join('\n');
}

function getFilterOperator(type: CustomFieldType): string {
    switch (type) {
        case 'datetime':
            return 'DateOperators';
        case 'string':
        case 'localeString':
            return 'StringOperators';
        case 'boolean':
            return 'BooleanOperators';
        case 'int':
        case 'float':
            return 'NumberOperators';
        default:
            assertNever(type);
    }
    return 'String';
}

function getGraphQlType(type: CustomFieldType): GraphQLFieldType {
    switch (type) {
        case 'string':
        case 'localeString':
            return 'String';
        case 'datetime':
            return 'DateTime';
        case 'boolean':
            return 'Boolean';
        case 'int':
            return 'Int';
        case 'float':
            return 'Float';
        default:
            assertNever(type);
    }
    return 'String';
}
