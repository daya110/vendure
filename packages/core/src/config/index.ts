export * from './asset-naming-strategy/asset-naming-strategy';
export * from './asset-naming-strategy/default-asset-naming-strategy';
export * from './asset-preview-strategy/asset-preview-strategy';
export * from './asset-storage-strategy/asset-storage-strategy';
export * from './asset-storage-strategy/local-asset-storage-strategy';
export * from './entity-id-strategy/auto-increment-id-strategy';
export * from './config.service';
export * from './entity-id-strategy/entity-id-strategy';
export * from './entity-id-strategy/uuid-id-strategy';
export * from './order-merge-strategy/order-merge-strategy';
export * from './logger/vendure-logger';
export * from './logger/default-logger';
export * from './logger/noop-logger';
export * from './payment-method/example-payment-method-config';
export * from './payment-method/payment-method-handler';
export * from './promotion/default-promotion-actions';
export * from './promotion/default-promotion-conditions';
export * from './promotion/promotion-action';
export * from './promotion/promotion-condition';
export * from './shipping-method/default-shipping-calculator';
export * from './shipping-method/default-shipping-eligibility-checker';
export * from './shipping-method/shipping-calculator';
export * from './shipping-method/shipping-eligibility-checker';
export * from './vendure-config';
export * from './merge-config';
