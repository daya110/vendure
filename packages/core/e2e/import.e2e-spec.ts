import gql from 'graphql-tag';
import path from 'path';

import { TEST_SETUP_TIMEOUT_MS } from './config/test-config';
import { TestAdminClient } from './test-client';
import { TestServer } from './test-server';

describe('Import resolver', () => {
    const client = new TestAdminClient();
    const server = new TestServer();

    beforeAll(async () => {
        const token = await server.init(
            {
                productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-empty.csv'),
                customerCount: 0,
            },
            {
                customFields: {
                    Product: [{ type: 'string', name: 'pageType' }],
                    ProductVariant: [{ type: 'int', name: 'weight' }],
                },
            },
        );
        await client.init();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('imports products', async () => {
        // TODO: waste a few more hours actually fixing this for real
        // Forgive me this abomination of a work-around.
        // On the inital run (as in CI), when the sqlite db has just been populated,
        // this test will fail due to an "out of memory" exception originating from
        // SqljsQueryRunner.ts:79:22, which is part of the findOne() operation on the
        // Session repository called from the AuthService.validateSession() method.
        // After several hours of fruitless hunting, I did what any desperate JavaScript
        // developer would do, and threw in a setTimeout. Which of course "works"...
        const timeout = process.env.CI ? 2000 : 1000;
        await new Promise(resolve => {
            setTimeout(resolve, timeout);
        });

        const csvFile = path.join(__dirname, 'fixtures', 'product-import.csv');
        const result = await client.importProducts(csvFile);

        expect(result.importProducts.errors).toEqual([
            'Invalid Record Length: header length is 16, got 1 on line 8',
        ]);
        expect(result.importProducts.imported).toBe(4);
        expect(result.importProducts.processed).toBe(4);

        const productResult = await client.query(
            gql`
                query GetProducts($options: ProductListOptions) {
                    products(options: $options) {
                        totalItems
                        items {
                            id
                            name
                            slug
                            description
                            featuredAsset {
                                id
                                name
                                preview
                                source
                            }
                            assets {
                                id
                                name
                                preview
                                source
                            }
                            optionGroups {
                                id
                                code
                                name
                            }
                            facetValues {
                                id
                                name
                                facet {
                                    id
                                    name
                                }
                            }
                            customFields {
                                pageType
                            }
                            variants {
                                id
                                name
                                sku
                                price
                                taxCategory {
                                    id
                                    name
                                }
                                options {
                                    id
                                    code
                                }
                                assets {
                                    id
                                    name
                                    preview
                                    source
                                }
                                featuredAsset {
                                    id
                                    name
                                    preview
                                    source
                                }
                                facetValues {
                                    id
                                    code
                                    name
                                    facet {
                                        id
                                        name
                                    }
                                }
                                stockOnHand
                                trackInventory
                                stockMovements {
                                    items {
                                        ... on StockMovement {
                                            id
                                            type
                                            quantity
                                        }
                                    }
                                }
                                customFields {
                                    weight
                                }
                            }
                        }
                    }
                }
            `,
            {
                options: {},
            },
        );

        expect(productResult.products.totalItems).toBe(4);
        expect(productResult.products.items).toMatchSnapshot();
    }, 20000);
});
