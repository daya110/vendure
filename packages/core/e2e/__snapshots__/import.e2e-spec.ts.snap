// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import resolver imports products 1`] = `
Array [
  Object {
    "assets": Array [
      Object {
        "id": "T_1",
        "name": "pps1.jpg",
        "preview": "test-url/test-assets/pps1__preview.jpg",
        "source": "test-url/test-assets/pps1.jpg",
      },
      Object {
        "id": "T_2",
        "name": "pps2.jpg",
        "preview": "test-url/test-assets/pps2__preview.jpg",
        "source": "test-url/test-assets/pps2.jpg",
      },
    ],
    "customFields": Object {
      "pageType": "default",
    },
    "description": "A great device for stretching paper.",
    "facetValues": Array [],
    "featuredAsset": Object {
      "id": "T_1",
      "name": "pps1.jpg",
      "preview": "test-url/test-assets/pps1__preview.jpg",
      "source": "test-url/test-assets/pps1.jpg",
    },
    "id": "T_1",
    "name": "Perfect Paper Stretcher",
    "optionGroups": Array [
      Object {
        "code": "perfect-paper-stretcher-size",
        "id": "T_1",
        "name": "size",
      },
    ],
    "slug": "perfect-paper-stretcher",
    "variants": Array [
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 100,
        },
        "facetValues": Array [
          Object {
            "code": "kb",
            "facet": Object {
              "id": "T_1",
              "name": "Brand",
            },
            "id": "T_1",
            "name": "KB",
          },
          Object {
            "code": "accessory",
            "facet": Object {
              "id": "T_2",
              "name": "Type",
            },
            "id": "T_2",
            "name": "Accessory",
          },
        ],
        "featuredAsset": null,
        "id": "T_1",
        "name": "Perfect Paper Stretcher Half Imperial",
        "options": Array [
          Object {
            "code": "half-imperial",
            "id": "T_1",
          },
        ],
        "price": 4530,
        "sku": "PPS12",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 100,
        },
        "facetValues": Array [
          Object {
            "code": "kb",
            "facet": Object {
              "id": "T_1",
              "name": "Brand",
            },
            "id": "T_1",
            "name": "KB",
          },
          Object {
            "code": "accessory",
            "facet": Object {
              "id": "T_2",
              "name": "Type",
            },
            "id": "T_2",
            "name": "Accessory",
          },
        ],
        "featuredAsset": null,
        "id": "T_2",
        "name": "Perfect Paper Stretcher Quarter Imperial",
        "options": Array [
          Object {
            "code": "quarter-imperial",
            "id": "T_2",
          },
        ],
        "price": 3250,
        "sku": "PPS14",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 100,
        },
        "facetValues": Array [
          Object {
            "code": "kb",
            "facet": Object {
              "id": "T_1",
              "name": "Brand",
            },
            "id": "T_1",
            "name": "KB",
          },
          Object {
            "code": "accessory",
            "facet": Object {
              "id": "T_2",
              "name": "Type",
            },
            "id": "T_2",
            "name": "Accessory",
          },
        ],
        "featuredAsset": null,
        "id": "T_3",
        "name": "Perfect Paper Stretcher Full Imperial",
        "options": Array [
          Object {
            "code": "full-imperial",
            "id": "T_3",
          },
        ],
        "price": 5950,
        "sku": "PPSF",
        "stockMovements": Object {
          "items": Array [
            Object {
              "id": "T_1",
              "quantity": -10,
              "type": "ADJUSTMENT",
            },
          ],
        },
        "stockOnHand": -10,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
    ],
  },
  Object {
    "assets": Array [],
    "customFields": Object {
      "pageType": "expanded",
    },
    "description": "Mabef description",
    "facetValues": Array [],
    "featuredAsset": null,
    "id": "T_2",
    "name": "Mabef M/02 Studio Easel",
    "optionGroups": Array [],
    "slug": "mabef-m02-studio-easel",
    "variants": Array [
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 300,
        },
        "facetValues": Array [
          Object {
            "code": "mabef",
            "facet": Object {
              "id": "T_1",
              "name": "Brand",
            },
            "id": "T_3",
            "name": "Mabef",
          },
          Object {
            "code": "easel",
            "facet": Object {
              "id": "T_2",
              "name": "Type",
            },
            "id": "T_4",
            "name": "Easel",
          },
        ],
        "featuredAsset": null,
        "id": "T_4",
        "name": "Mabef M/02 Studio Easel",
        "options": Array [],
        "price": 91070,
        "sku": "M02",
        "stockMovements": Object {
          "items": Array [
            Object {
              "id": "T_2",
              "quantity": 100,
              "type": "ADJUSTMENT",
            },
          ],
        },
        "stockOnHand": 100,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
    ],
  },
  Object {
    "assets": Array [],
    "customFields": Object {
      "pageType": "default",
    },
    "description": "Really mega pencils",
    "facetValues": Array [],
    "featuredAsset": null,
    "id": "T_3",
    "name": "Giotto Mega Pencils",
    "optionGroups": Array [
      Object {
        "code": "giotto-mega-pencils-box-size",
        "id": "T_2",
        "name": "box size",
      },
    ],
    "slug": "giotto-mega-pencils",
    "variants": Array [
      Object {
        "assets": Array [
          Object {
            "id": "T_3",
            "name": "box-of-8.jpg",
            "preview": "test-url/test-assets/box-of-8__preview.jpg",
            "source": "test-url/test-assets/box-of-8.jpg",
          },
        ],
        "customFields": Object {
          "weight": 200,
        },
        "facetValues": Array [
          Object {
            "code": "xmas-sale",
            "facet": Object {
              "id": "T_3",
              "name": "Collection",
            },
            "id": "T_5",
            "name": "Xmas Sale",
          },
        ],
        "featuredAsset": Object {
          "id": "T_3",
          "name": "box-of-8.jpg",
          "preview": "test-url/test-assets/box-of-8__preview.jpg",
          "source": "test-url/test-assets/box-of-8.jpg",
        },
        "id": "T_5",
        "name": "Giotto Mega Pencils Box of 8",
        "options": Array [
          Object {
            "code": "box-of-8",
            "id": "T_4",
          },
        ],
        "price": 416,
        "sku": "225400",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [
          Object {
            "id": "T_4",
            "name": "box-of-12.jpg",
            "preview": "test-url/test-assets/box-of-12__preview.jpg",
            "source": "test-url/test-assets/box-of-12.jpg",
          },
        ],
        "customFields": Object {
          "weight": 200,
        },
        "facetValues": Array [
          Object {
            "code": "xmas-sale",
            "facet": Object {
              "id": "T_3",
              "name": "Collection",
            },
            "id": "T_5",
            "name": "Xmas Sale",
          },
        ],
        "featuredAsset": Object {
          "id": "T_4",
          "name": "box-of-12.jpg",
          "preview": "test-url/test-assets/box-of-12__preview.jpg",
          "source": "test-url/test-assets/box-of-12.jpg",
        },
        "id": "T_6",
        "name": "Giotto Mega Pencils Box of 12",
        "options": Array [
          Object {
            "code": "box-of-12",
            "id": "T_5",
          },
        ],
        "price": 624,
        "sku": "225600",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
        "trackInventory": false,
      },
    ],
  },
  Object {
    "assets": Array [],
    "customFields": Object {
      "pageType": "default",
    },
    "description": "Keeps the paint off the clothes",
    "facetValues": Array [
      Object {
        "facet": Object {
          "id": "T_4",
          "name": "Material",
        },
        "id": "T_6",
        "name": "Denim",
      },
      Object {
        "facet": Object {
          "id": "T_3",
          "name": "Collection",
        },
        "id": "T_7",
        "name": "clothes",
      },
    ],
    "featuredAsset": null,
    "id": "T_4",
    "name": "Artists Smock",
    "optionGroups": Array [
      Object {
        "code": "artists-smock-size",
        "id": "T_3",
        "name": "size",
      },
      Object {
        "code": "artists-smock-colour",
        "id": "T_4",
        "name": "colour",
      },
    ],
    "slug": "artists-smock",
    "variants": Array [
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 500,
        },
        "facetValues": Array [],
        "featuredAsset": null,
        "id": "T_7",
        "name": "Artists Smock small beige",
        "options": Array [
          Object {
            "code": "small",
            "id": "T_6",
          },
          Object {
            "code": "beige",
            "id": "T_8",
          },
        ],
        "price": 1199,
        "sku": "10112",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 500,
        },
        "facetValues": Array [],
        "featuredAsset": null,
        "id": "T_8",
        "name": "Artists Smock large beige",
        "options": Array [
          Object {
            "code": "large",
            "id": "T_7",
          },
          Object {
            "code": "beige",
            "id": "T_8",
          },
        ],
        "price": 1199,
        "sku": "10113",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 500,
        },
        "facetValues": Array [],
        "featuredAsset": null,
        "id": "T_9",
        "name": "Artists Smock small navy",
        "options": Array [
          Object {
            "code": "small",
            "id": "T_6",
          },
          Object {
            "code": "navy",
            "id": "T_9",
          },
        ],
        "price": 1199,
        "sku": "10114",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
        "trackInventory": false,
      },
      Object {
        "assets": Array [],
        "customFields": Object {
          "weight": 500,
        },
        "facetValues": Array [],
        "featuredAsset": null,
        "id": "T_10",
        "name": "Artists Smock large navy",
        "options": Array [
          Object {
            "code": "large",
            "id": "T_7",
          },
          Object {
            "code": "navy",
            "id": "T_9",
          },
        ],
        "price": 1199,
        "sku": "10115",
        "stockMovements": Object {
          "items": Array [],
        },
        "stockOnHand": 0,
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
        "trackInventory": false,
      },
    ],
  },
]
`;
